### PR TITLE
fix iconv compile error on windows[NOT TO BE MERGED]

### PR DESCRIPTION
**Pull request by xiaoyifang**
this PR was extracted from #1400 ,has not tested against webkit goldendict  as I have no webkit environment and not bother to configure one

remained here only for a reference.

not aimed to be merged,but serve as a alternative temp way for someone who have encountered the same errors as me, to bypass the  execution error such as 'bad allocation ' or https://github.com/goldendict/goldendict/issues/1322 etc .

### CODE REVIEW

File: wstring.hh [hunk 1]
```diff
@@ -55,8 +55,8 @@ namespace gd
 
   #else
 
-  typedef wchar_t wchar;
-  using std::wstring;
+  typedef char32_t wchar;
+  typedef std::u32string wstring;
```
Comment (nonwill) [inline wstring.hh:59]: Changes to wstring.hh is not releated to the TOPIC of this pull request!！
Please make sure you are realy ready to make the request easily acceptable just as you do in #1447.

**Review by nonwill — CHANGES_REQUESTED** [review]

File: wstring.hh (hunk 1 above)
Comment (xiaoyifang) [reply wstring.hh:59]: iconv changed to qstream .   u32string was an easiler choice.
QString::toStdU32String()

Comment (nonwill) [reply wstring.hh:59]: >u32string was an easiler choice. QString::toStdU32String()

Yes, it's off this TOPIC.

File: dsl_details.cc [hunk 2]
```diff
@@ -1108,6 +1136,31 @@ char const * DslIconv::getEncodingNameFor( DslEncoding e )
...
+	case Windows1251:
+		
+	case Details::Utf8:
+		
+	case Windows1250:
+	default:
+        lineFeedLength = 1;
+        lineFeed = new char[1] {0x0A};
+	}
+}
```
Comment (nonwill) [inline dsl_details.cc:1162]: Format of your changes(codes) is not consistent with the original code!

**Review by JingDict — CHANGES_REQUESTED** [review]

**Comment by nonwill** [comment]
> fix iconv compile error on windows

In addition to the one side of story in #1322, can you elaborate clearly on the error? Because I (and also @Abs62 ?) have never run into this problem.

**Comment by nonwill** [comment]
> should wait until #1447 been merged.

If the pull request is related to single issue, how and why does it depend on #1447?

**Comment by nonwill** [comment]
>fix iconv compile error on windows

In addition to the one side of story in https://github.com/goldendict/goldendict/issues/1322, can you elaborate clearly on the error? Because I (and also @Abs62 ?) have never run into this problem.

>should wait until #1447 been merged.

If the pull request is related to single issue, how and why does it depend on #1447?

It just looks like you have a clear and strong motive to push your #1477, but not really to identify problems, analyze them and then solve them.

**Comment by nonwill** [comment]
>fix iconv compile error on windows

In addition to the one side of story in https://github.com/goldendict/goldendict/issues/1322, can you elaborate clearly on the error? Because I (and also @Abs62 ?) have never run into problems.

>should wait until #1447 been merged.

If the pull request is related to single issue, how and why does it depend on #1447?

It just looks like you have a clear and strong motive to push your #1477, but not really to identify problems, analyze them and then solve them.

**Comment by nonwill** [comment]
>fix iconv compile error on windows

In addition to the one side of story in https://github.com/goldendict/goldendict/issues/1322, can you elaborate clearly on the error? Because I (and also @Abs62 ?) have never run into problems.

>should wait until #1447 been merged.

If the pull request is related to single issue, how and why does it depend on #1447?

>If this is really a issue ,it is about ffmpeg build problems on windows which clearly not included in this PR.
>_Originally posted by @xiaoyifang in https://github.com/goldendict/goldendict/issues/1447#issuecomment-1028626080_

It just looks like you have a clear and strong motive to push your #1447 , but not really to identify problems, analyze them and then solve them.

**Comment by nonwill** [comment]
> qstream peformance

@xiaoyifang : Even if there is no error as you described, can you give some performance comparison(better with charts) to show how much of the difference between iconv and qstream(actually it should be QTextCodec) really is?

**Comment by vedgy** [comment]
In order to figure out if there is an issue at all, and if yes, what it is, how about comparing the steps to build iconv on Windows?
@nonwill, @xiaoyifang, could you please post your detailed build steps? (including the version of the library and where it was downloaded from)

**Comment by xiaoyifang** [comment]
> If the problem is not in the build steps, then perhaps a buggy DSL dictionary is used? @xiaoyifang, you could upload a problematic dictionary and ask others to load it in Debug mode on Windows.

I think dsl is ok. I just used the iconv provided [here](https://github.com/goldendict/goldendict#:~:text=GoldenDict_libs_VS2015_x64_v4.7z%20%2D%20for%20MS%20Visual%20Studio%202015%2C%2064%20bit)

**Comment by nonwill** [comment]
> detailed build steps?

Windows build: https://github.com/pffang/libiconv-for-Windows with msvc2019 project files,
Or we can just convert the prj file to pro, and then build it with qmake tools.

**Comment by vedgy** [comment]
@xiaoyifang, the link you have posted contains `text=GoldenDict_libs_VS2015_x64_v4.7z - for MS Visual Studio 2015%2C 64 bit`. Could the issue be in the incompatibility of Visual Studio 2015 and 2019 versions? Can't you compile this library manually?

**Comment by xiaoyifang** [comment]
> If the pull request is related to single issue, how and why does it depend on #1447?

not dependent on #1447 ,as I have no webkit environment to verify. which I have said many times. and you know the reason .
[comments](https://github.com/goldendict/goldendict/pull/1447#:~:text=if%20a%20sperate,be%20on%20webengine.)

**Comment by xiaoyifang** [comment]
> @xiaoyifang, the link you have posted contains `text=GoldenDict_libs_VS2015_x64_v4.7z - for MS Visual Studio 2015%2C 64 bit`. Could the issue be in the incompatibility of Visual Studio 2015 and 2019 versions? Can't you compile this library manually?

maybe true, so I think I'll just close this.

**Comment by vedgy** [comment]
@xiaoyifang, you wrote in #1447:

> > So how about creating a separate small pull request with your Windows iconv fix?
> 
> without iconv changes goldendict can not work with dsl and mdx,though great possibility I have used the mismatched iconv .
> I can remove it ,but I can never verify the changes (without it) would be correct.

Please try to compile iconv correctly to be able to verify that the QtWebEngine porting works as intended. Post your results here or under #1322 as they could be useful to other people who compile GoldenDict on Windows.

**Comment by xiaoyifang** [comment]
I agree with you that iconv and qtwebengine are two different issues.

as I have not change iconv code in #1322  ,and you do not use it on windows ,I think you can verify the qwebengine's function in Linux etc.

I do not want to waste my time on compile iconv.

**Comment by xiaoyifang** [comment]
Finally I decided to open it again ,not aimed to be merged,but as a alternative temp way to bypass the compile or execute error.such as 'bad allocation 'etc .

**Comment by xiaoyifang** [comment]
> could you please post your detailed build steps? (including the version of the library and where it was downloaded from)

I have tried all the following ways, all failed.
* use goldendict bundled iconv.lib in winlib folder
* copy from other software's folders
* https://github.com/pffang/libiconv-for-Windows  ,shared release.
* vcpkg install iconv  and copied to gd

**Comment by vedgy** [comment]
Is any of the iconv binaries you have tried compiled with the same compiler bit-ness and version as you compile GoldenDict?

**Comment by xiaoyifang** [comment]
> * https://github.com/pffang/libiconv-for-Windows  ,shared release.
> * vcpkg install iconv  and copied to gd

"https://github.com/pffang/libiconv-for-Windows ,shared release."
this was compiled with windows64+visual studio 2019+release 
"vcpkg install iconv and copied to gd"
 compiled with vcpkg x64-windows-release triplet.

**Pull request closed without merging**